contracts/Orchestrator.sol

```diff
@@ -1,53 +1,72 @@
-pragma solidity 0.7.6;
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.2;
 
-import "./_external/Ownable.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {BytesLib} from "./_external/BytesLib.sol";
+import {StringUtils} from "./_external/StringUtils.sol";
 
-import "./UFragmentsPolicy.sol";
+interface IUFragmentsPolicy {
+    function rebase() external;
+}
 
 /**
  * @title Orchestrator
  * @notice The orchestrator is the main entry point for rebase operations. It coordinates the policy
  * actions with external consumers.
  */
 contract Orchestrator is Ownable {
+    using BytesLib for bytes;
+    using StringUtils for uint16;
+
+    // Reference to the Ampleforth Policy
+    address public policy;
+
     struct Transaction {
         bool enabled;
+        bool critical;
         address destination;
         bytes data;
     }
 
     // Stable ordering is not guaranteed.
     Transaction[] public transactions;
 
-    UFragmentsPolicy public policy;
+    // events
+    event TransactionFailed(uint16 index);
 
     /**
      * @param policy_ Address of the UFragments policy.
      */
-    constructor(address policy_) public {
-        Ownable.initialize(msg.sender);
-        policy = UFragmentsPolicy(policy_);
+    constructor(address policy_) {
+        policy = policy_;
     }
 
     /**
      * @notice Main entry point to initiate a rebase operation.
      *         The Orchestrator calls rebase on the policy and notifies downstream applications.
      *         Contracts are guarded from calling, to avoid flash loan attacks on liquidity
      *         providers.
-     *         If a transaction in the transaction list fails, Orchestrator will stop execution
-     *         and revert to prevent a gas underprice attack.
+     *         If a transaction marked 'critical' in the transaction list fails,
+     *         Orchestrator will stop execution and revert.
      */
     function rebase() external {
         require(msg.sender == tx.origin); // solhint-disable-line avoid-tx-origin
 
-        policy.rebase();
+        IUFragmentsPolicy(policy).rebase();
 
-        for (uint256 i = 0; i < transactions.length; i++) {
+        for (uint16 i = 0; i < transactions.length; i++) {
             Transaction storage t = transactions[i];
             if (t.enabled) {
-                (bool result, ) = t.destination.call(t.data);
-                if (!result) {
-                    revert("Transaction Failed");
+                (bool success, bytes memory reason) = t.destination.call(t.data);
+
+                // Critical transaction failed, revert with message
+                if (!success && t.critical) {
+                    revert(buildRevertReason(i, reason));
+                }
+
+                // Non-Critical transaction failed, log error and continue
+                if (!success) {
+                    emit TransactionFailed(i);
```

Comment:
I was thinking we could also include an error message in this log message. But I'm thinking maybe not, now. A non-critical TX might exist for awhile if we're waiting on an external system to fix itself. And we don't want to risk hitting gas limits unnecessarily.

Critical ones will have to get fixed fast or removed, so I think it's good to have there.

Reply:
We could add it for consistency. Dont think we risk hitting gas limits anytime soon ..